In [26]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

In [27]:
data=pd.read_csv("/content/StudentPerformanceFactors.csv")

In [28]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6607 entries, 0 to 6606
Data columns (total 20 columns):
 #   Column                      Non-Null Count  Dtype 
---  ------                      --------------  ----- 
 0   Hours_Studied               6607 non-null   int64 
 1   Attendance                  6607 non-null   int64 
 2   Parental_Involvement        6607 non-null   object
 3   Access_to_Resources         6607 non-null   object
 4   Extracurricular_Activities  6607 non-null   object
 5   Sleep_Hours                 6607 non-null   int64 
 6   Previous_Scores             6607 non-null   int64 
 7   Motivation_Level            6607 non-null   object
 8   Internet_Access             6607 non-null   object
 9   Tutoring_Sessions           6607 non-null   int64 
 10  Family_Income               6607 non-null   object
 11  Teacher_Quality             6529 non-null   object
 12  School_Type                 6607 non-null   object
 13  Peer_Influence              6607 non-null   obje

In [11]:
from sklearn.preprocessing import LabelEncoder
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import accuracy_score

In [37]:
from sklearn.preprocessing import LabelEncoder

le1 = LabelEncoder()
le2 = LabelEncoder()

data["Motivation_Level"] = le1.fit_transform(
    data["Motivation_Level"]
)

data["Parental_Involvement"] = le2.fit_transform(
    data["Parental_Involvement"]
)

In [38]:
print(data["Motivation_Level"].unique())
print(data["Parental_Involvement"].unique())

[1 2 0]
[1 2 0]


In [44]:
from sklearn.linear_model import LinearRegression

model = LinearRegression()

model.fit(x_train,y_train)

LinearRegression()

In [47]:
predictions=model.predict(x_test)
print(predictions[0])

67.03154558048398


In [48]:
print(y_test.iloc[0])

65


In [50]:
COMPARISON=pd.DataFrame({"Actual Values":y_test,"Predicted Values":predictions})
COMPARISON

,Actual Values,Predicted Values
743,65,67.031546
5551,65,67.476789
3442,71,68.245095
6571,64,64.609763
4204,66,67.525348
...,...,...
4709,69,67.702069
3664,72,69.708987
5231,66,68.355077
1773,63,67.653510


## First Prediction

Predicted Score: 67.03

Actual Score: 65

Difference: 2.03

Observation:
The model was able to generate a prediction close to the actual exam score, showing that it successfully learned patterns from the training data.

## Key Takeaways

- Linear Regression is used to predict continuous numerical values.
- Features (X) are the input variables used for prediction.
- Target (y) is the output variable we want to predict.
- The model learns patterns using fit().
- The model makes predictions using predict().
- Data preprocessing is often as important as model building.
- Real-world datasets may contain categorical values that need to be encoded before training.